x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/nio/SSLDriver.java

```diff
@@ -106,22 +109,24 @@ public boolean isHandshaking() {
         return currentMode.isHandshake();
     }
 
-    public ByteBuffer getNetworkReadBuffer() {
-        return networkReadBuffer;
-    }
-
     public SSLOutboundBuffer getOutboundBuffer() {
         return outboundBuffer;
     }
 
-    public void read(InboundChannelBuffer buffer) throws SSLException {
-        Mode modePriorToRead;
-        do {
-            modePriorToRead = currentMode;
-            currentMode.read(buffer);
-            // If we switched modes we want to read again as there might be unhandled bytes that need to be
-            // handled by the new mode.
-        } while (modePriorToRead != currentMode);
+    public void read(InboundChannelBuffer encryptedBuffer, InboundChannelBuffer applicationBuffer) throws SSLException {
+        networkReadPage = pageAllocator.apply(packetSize);
+        try {
+            Mode modePriorToRead;
+            do {
+                modePriorToRead = currentMode;
+                currentMode.read(encryptedBuffer, applicationBuffer);
+                // If we switched modes we want to read again as there might be unhandled bytes that need to be
+                // handled by the new mode.
+            } while (modePriorToRead != currentMode);
```

Comment:
how can this be  `false` ever? I mean we never assign `modePriorToRead` again. The only explanation is that currentMode.read can have a sideeffect. If that is true can you leave some more verbose comment here. it took me quite a while to understand this.

Reply:
> The only explanation is that currentMode.read can have a sideeffect

This is indeed the case. +1 to add a comment about this